First pass at linalg_ext.scan op

harsh-nod opened the pull request:
This pass adds the first version of the scan op in
the linalg_ext dialect. Currently, the op requires specifying
the dimension, whether the scan is inclusive or not as well
as body specifying the operator for scan.

The current patch adds code to lower the op to loops
using the naive sequential implementation. The tiled implementation
is borrowed from the sort op.

Finally, some unit tests have also been added to test the passes.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -247,6 +247,60 @@ def IREELinalgExt_FftOp : IREELinalgExt_Op<"fft", [
+
+  let arguments = (ins Variadic<AnyType>:$inputs,
+                       Variadic<AnyShaped>:$outputs,
+                       OptionalAttr<I64Attr>:$dimension,
```
Instead of making this optional, I'd rather make then required always, and have builder methods that set defaults for these. You can do something like

```
let builders = [
  OpBuilder<(ins "ValueRange":$inputs, "ValueRange":$outputs,
    CArg<"int64_t", "0">:$dimension, CArg<"bool", "true">:$inclusive)>
];
```

MaheshRavishankar replied:
The defaults are just convenience methods for builders. It will just generate an extra build  method that will set the value of dimension to 0 and inclusive to true if you call that method.

harsh-nod replied:
I see, makes sene thanks!

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -247,6 +247,60 @@ def IREELinalgExt_FftOp : IREELinalgExt_Op<"fft", [
+    int64_t getOperandRank() {
+      return getOperandType().getRank();
+    }
+    uint64_t getScanDimension() {
```
If you make this required always, then you can drop these methods `getScanDimension()` can just be `dimension()` and `isInclusive()` will just be `inclusive()`

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,142 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  Value source = input();
+  for (auto dim : llvm::seq<int64_t>(0, operandRank)) {
```
Wont the number of iterations for the scan dimension change if this an inclusive or exclusive scan? Depends on how you would design this op and how it lowers to loops.

MaheshRavishankar replied:
You are right. My bad. I was Thinking that for the exclusive scan the size of the result is 1 greater than the size of the input to hold the value of the last element. Seems like thats not the case. So this is good.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,142 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verifyScanOp(ScanOp op) {
+  if (op.inputs().size() != 1) {
```
I think the shape of the output w.r.t input needs to be checked as well for the inclusive and exclusive cases (only when the shapes are static, dynamic case you cant really check at this level)

MaheshRavishankar commented on llvm-external-projects/iree-dialects/test/iree_linalgext/convert_to_loops.mlir:
```diff
@@ -505,3 +505,90 @@ func @reverse_dim_0(%arg0: memref<?x?xi32>, %arg1: memref<?x?xi32>) {
+// CHECK-DAG:     %[[C1:.+]] = arith.constant 1 : index
+// CHECK:         scf.for %[[ARG1:.+]] = %[[C0]] to %[[C128]] step %[[C1]]
+// CHECK:           %[[COND:.+]] = arith.cmpi eq, %[[ARG1]], %[[C0]] : index
+// CHECK:           scf.if %[[COND]] {
```
I'd avoid the `if` if possible and just peel the loop. 

So inclusive would be
```
output[0] = input[0]
for i = 1 to ub step 1 {
   output[i] = output[i-1] + input[i]
}
```

Exclusive would be
```
output[0] = 0
for i = 0 to ub step 1 {
  output[i+1] = output[i] + input[i];
}
```

MaheshRavishankar replied:
Good point. For now lets just stay with the `scf.if` version you have then. This is only required for last step and hopefully the classic loop rotation transformation in LLVM will do the right thing.

harsh-nod replied:
Sounds good. thanks!

MaheshRavishankar reviewed (changes requested):
Thanks for pushing on this. First set of comments.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -247,6 +247,60 @@ def IREELinalgExt_FftOp : IREELinalgExt_Op<"fft", [
+  }];
+
+  let extraClassDeclaration = extraLinalgExtOpClassDeclaration # [{
+    Value input() {
```
You shouldnt have to worry about the `AttrSizedOperandSegments` trait. If you make them non-variadic, then the generated op definition will not need that trait. So you should be able to say just 
```
 let arguments = (ins AnyShaped:$input,
                            AnyShaped:$output,
                            I64Attr:$dimension,
                            BoolAttr:$inclusive
  );
```

harsh-nod replied:
So the problem is that the IREELinalgExt_ScanOp is a type of an IREELinalgExt_Op which in turn is a type of IREELinalgExt_PureOp. The pure op has that trait which means all ext ops have that trait. From https://github.com/llvm/llvm-project/blob/main/mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp#L963, if the op has that trait and has less than 2 variadic args, then tablegen complains. I tried doing what you suggested above and got the error

```
error: op must have at least two variadic operands to use 'AttrSizedOperandSegments' trait
def IREELinalgExt_ScanOp : IREELinalgExt_Op<"scan", [
^
```

Am I missing something here or doing something wrong?

harsh-nod replied:
I see that makes sense. Okay I have the following checks in verifyScanOp 
```C++
if (op.getNumInputs() != 1) {
  return op.emitOpError("expected one input operands");
}
if (op.getNumOutputs() != 1) {
  return op.emitOpError("expected one output operand");
}
```

Is that what you were looking for?

MaheshRavishankar replied:
Great! thanks!

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,161 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+                                          ArrayRef<OpFoldResult> offsets,
+                                          ArrayRef<OpFoldResult> sizes,
+                                          SmallVectorImpl<Value> &results) {
+  assert(outputs.size() == this->outputs().size());
```
I see that you have the `getTiledImplementation` implemented, but there is not test for the scan actually being tiled. You can add a test to `tiling.mlir`. That should pick up the implementation her and generate the tiled code. If you want to not do that right now, just return `nullptr` here and implement the tiling on a future PR.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -247,6 +247,52 @@ def IREELinalgExt_FftOp : IREELinalgExt_Op<"fft", [
+  let assemblyFormat = [{
+    `dimension` `(` $dimension `)`
+    `inclusive` `(` $inclusive `)`
+    attr-dict (`ins` `(` $inputs^ `:` type($inputs) `)`)?
```
I think for the `scan` inputs cannot be optional. So this should just be 
```
attr-dict `ins` `(` $inputs `:` type($inputs) `)`
```

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,161 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+          b.create<memref::StoreOp>(loc, value, output(), indices);
+        } else {
+          auto elementType = getOperandType().getElementType();
+          Value identity = b.create<arith::ConstantOp>(
```
This makes me think the `identity` here should also be an operand to the op. Here using zero assumes that the operation is a sum.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,161 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+
+  auto &srcBlock = region().front();
+  Region &region = scfIf.getElseRegion();
+  BlockAndValueMapping bvm;
```
Nice!

MaheshRavishankar reviewed (changes requested):
I think some comments are still not addressed. I just saw a notification, but wasnt sure if I missed some updates.

harsh-nod commented:
> I think some comments are still not addressed. I just saw a notification, but wasnt sure if I missed some updates.

I think I have addressed all your comments now. Let me know if I missed something.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/test/iree_linalgext/tiling.mlir:
```diff
@@ -1206,3 +1206,43 @@ func @pad_tensor(%arg0 : tensor<?x?xf32>, %arg1 : index, %arg2 : index,
+
+// -----
+
+func @scan_1d(%0: tensor<128xi32>) -> tensor<128xi32> {
```
This didnt actually tile the operation did it? You need to add `{__internal_linalg_transform__ = "tiling_input"}` to the op (its an extra attribute to tell the pass to tile the operation).

harsh-nod replied:
Thanks! Yes, I missed that. When I do that, and modify the IR to be
```mlir
  func @scan_1d(%0: tensor<128xi32>) -> tensor<128xi32> {
    %c0 = arith.constant 0 : i32
    %1 = linalg.init_tensor [128] : tensor<128xi32>
    %2 = iree_linalg_ext.scan
      dimension(0) inclusive(true)
      {__internal_linalg_transform__ = "tiling_input"}
      identity(%c0 : i32)
      ins(%0 : tensor<128xi32>) outs(%1 : tensor<128xi32>) {
      ^bb0(%arg0 : i32, %arg1 : i32):
        %sum = arith.addi %arg0, %arg1 : i32
        iree_linalg_ext.yield %sum : i32
    } -> tensor<128xi32>
    return %2 : tensor<128xi32>
  }
```

Then, I get the following error
```
error: 'iree_linalg_ext.scan' op unimplemented tiling of non-parallel loop iterator type
```

But I thought I provided a getTiledImplementation, so not sure why I would be hitting this error. Any ideas on what could be causing this?

Thanks!

harsh-nod replied:
Instead of creating a new set of tiling options, I just re-used "outer_reduce_input". Let me know what you think.

MaheshRavishankar replied:
Sorry, I missed this. What you did is right. You cannot specify a tiling value for a reduced iterator. Thats the error. Its just a matter of specifying the correct configuration for tiling. When plugged into IREE, we will have to specify a way to handle this if needed, but for testing what you have is correct.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -767,6 +767,178 @@ Operation *FftOp::getTiledImplementation(OpBuilder &builder, ValueRange outputs,
+  SmallVector<unsigned> partitionableLoops(range.begin(), range.end());
+  partitionableLoops.erase(std::next(partitionableLoops.begin(), dimension()));
+  if (partitionableLoops.size() > maxNumParallelDims) {
+    partitionableLoops.erase(
```
I think for now you can drop lines 835-840. We can revisit this later. I think the caller already handles this properly (or at least should). The argument here is just to allow the ops to make a choice if there are more parallel dims than whats supported.

harsh-nod replied:
Sounds good. Will remove those lines for now.

MaheshRavishankar reviewed (approved):
Nice! This looks good.

harsh-nod commented:
> Nice! This looks good.

Thanks! Do you want me to squash all my commits into 1 commit before pushing?

MaheshRavishankar commented:
> > Nice! This looks good.
> 
> Thanks! Do you want me to squash all my commits into 1 commit before pushing?

You need to squash and merge the commit.
